File: Robotics Kinematics and Dynamics Course/Chapter12/Lesson5/Chapter12_Lesson5.py
import numpy as np

class TwoLinkParams:
    def __init__(self, m1, m2, l1, c1, c2, I1, I2, g=9.81):
        self.m1 = m1
        self.m2 = m2
        self.l1 = l1
        self.c1 = c1
        self.c2 = c2
        self.I1 = I1
        self.I2 = I2
        self.g  = g

def M_lagrange_2R(q, p):
    q1, q2 = q
    a = p.I1 + p.I2 + p.m1 * p.c1**2 + p.m2 * (p.l1**2 + p.c2**2)
    b = p.m2 * p.l1 * p.c2
    d = p.I2 + p.m2 * p.c2**2
    c2 = np.cos(q2)
    M11 = a + 2.0 * b * c2
    M12 = d + b * c2
    M22 = d
    return np.array([[M11, M12],
                     [M12, M22]], dtype=float)

def h_lagrange_2R(q, dq, p):
    q1, q2 = q
    dq1, dq2 = dq
    b = p.m2 * p.l1 * p.c2
    s2 = np.sin(q2)
    h1 = -b * s2 * (2.0 * dq1 * dq2 + dq2**2)
    h2 =  b * s2 * dq1**2
    return np.array([h1, h2], dtype=float)

def g_lagrange_2R(q, p):
    q1, q2 = q
    g1 = p.g * (p.c1 * p.m1 * np.cos(q1)
                + p.m2 * (p.l1 * np.cos(q1) + p.c2 * np.cos(q1 + q2)))
    g2 = p.g * (p.m2 * p.c2 * np.cos(q1 + q2))
    return np.array([g1, g2], dtype=float)

def tau_lagrange_2R(q, dq, ddq, p):
    M = M_lagrange_2R(q, p)
    h = h_lagrange_2R(q, dq, p)
    g = g_lagrange_2R(q, p)
    return M.dot(ddq) + h + g

# ----------------------------------------------------------------------
# Newton-Euler interface: you should plug in your own implementation
# from Chapter 12, Lesson 4. Here we declare the expected signature.
# ----------------------------------------------------------------------
def tau_newton_euler_2R(q, dq, ddq, p, g_base=None):
    """
    Placeholder: call your validated Newton-Euler implementation here.
    Arguments:
        q, dq, ddq: arrays of shape (2,)
        p: TwoLinkParams
        g_base: 3D gravity vector in base frame (e.g. [0, -p.g, 0])
    Returns:
        tau_NE: array of shape (2,)
    """
    raise NotImplementedError("Connect to your Newton-Euler implementation.")

# ----------------------------------------------------------------------
# Reconstruction of M_NE(q) and g_NE(q) using the strategy of Section 5
# ----------------------------------------------------------------------
def reconstruct_M_NE(q, p):
    n = 2
    M = np.zeros((n, n), dtype=float)
    dq = np.zeros(n)
    g_base = np.zeros(3)  # zero gravity
    for j in range(n):
        ddq = np.zeros(n)
        ddq[j] = 1.0
        tau = tau_newton_euler_2R(q, dq, ddq, p, g_base=g_base)
        M[:, j] = tau
    return M

def reconstruct_g_NE(q, p):
    n = 2
    ddq = np.zeros(n)
    dq = np.zeros(n)
    # gravity pointing along negative y-axis
    g_base = np.array([0.0, -p.g, 0.0])
    tau = tau_newton_euler_2R(q, dq, ddq, p, g_base=g_base)
    return tau

def monte_carlo_validation(num_samples=100):
    p = TwoLinkParams(m1=2.0, m2=1.0,
                      l1=1.0, c1=0.5, c2=0.5,
                      I1=0.2, I2=0.1)
    max_err = 0.0
    for k in range(num_samples):
        q  = np.array([np.random.uniform(-np.pi, np.pi),
                       np.random.uniform(-np.pi, np.pi)])
        dq = np.array([np.random.uniform(-1.0, 1.0),
                       np.random.uniform(-1.0, 1.0)])
        ddq = np.array([np.random.uniform(-2.0, 2.0),
                        np.random.uniform(-2.0, 2.0)])

        tau_L  = tau_lagrange_2R(q, dq, ddq, p)
        tau_NE = tau_newton_euler_2R(q, dq, ddq, p,
                                     g_base=np.array([0.0, -p.g, 0.0]))
        err = np.linalg.norm(tau_L - tau_NE, ord=np.inf)
        max_err = max(max_err, err)

    print("Max absolute torque error over", num_samples, "samples:", max_err)

    # Single configuration structural test
    q0 = np.array([0.3, -0.7])
    M_L = M_lagrange_2R(q0, p)
    g_L = g_lagrange_2R(q0, p)
    M_NE = reconstruct_M_NE(q0, p)
    g_NE = reconstruct_g_NE(q0, p)

    print("M_L(q0) =\n", M_L)
    print("M_NE(q0) =\n", M_NE)
    print("M_L - M_NE =\n", M_L - M_NE)
    print("g_L(q0) =", g_L)
    print("g_NE(q0) =", g_NE)
    print("g_L - g_NE =", g_L - g_NE)

if __name__ == "__main__":
    # Once tau_newton_euler_2R is connected, this will
    # provide a strong regression test of your implementation.
    monte_carlo_validation()
      

File: Robotics Kinematics and Dynamics Course/Chapter12/Lesson5/Chapter12_Lesson5.cpp
#include <iostream>
#include <cmath>
#include <Eigen/Dense>

struct TwoLinkParams {
    double m1, m2;
    double l1, c1, c2;
    double I1, I2;
    double g;
};

Eigen::Matrix2d M_lagrange_2R(const Eigen::Vector2d& q,
                              const TwoLinkParams& p) {
    double q2 = q(1);
    double a = p.I1 + p.I2 + p.m1 * p.c1 * p.c1
               + p.m2 * (p.l1 * p.l1 + p.c2 * p.c2);
    double b = p.m2 * p.l1 * p.c2;
    double d = p.I2 + p.m2 * p.c2 * p.c2;
    double c2 = std::cos(q2);

    Eigen::Matrix2d M;
    M(0,0) = a + 2.0 * b * c2;
    M(0,1) = d + b * c2;
    M(1,0) = M(0,1);
    M(1,1) = d;
    return M;
}

Eigen::Vector2d h_lagrange_2R(const Eigen::Vector2d& q,
                              const Eigen::Vector2d& dq,
                              const TwoLinkParams& p) {
    double q2  = q(1);
    double dq1 = dq(0);
    double dq2 = dq(1);
    double b   = p.m2 * p.l1 * p.c2;
    double s2  = std::sin(q2);

    Eigen::Vector2d h;
    h(0) = -b * s2 * (2.0 * dq1 * dq2 + dq2 * dq2);
    h(1) =  b * s2 * dq1 * dq1;
    return h;
}

Eigen::Vector2d g_lagrange_2R(const Eigen::Vector2d& q,
                              const TwoLinkParams& p) {
    double q1 = q(0);
    double q2 = q(1);

    double g1 = p.g * (p.c1 * p.m1 * std::cos(q1)
                       + p.m2 * (p.l1 * std::cos(q1)
                                 + p.c2 * std::cos(q1 + q2)));
    double g2 = p.g * (p.m2 * p.c2 * std::cos(q1 + q2));

    Eigen::Vector2d g;
    g(0) = g1;
    g(1) = g2;
    return g;
}

Eigen::Vector2d tau_lagrange_2R(const Eigen::Vector2d& q,
                                const Eigen::Vector2d& dq,
                                const Eigen::Vector2d& ddq,
                                const TwoLinkParams& p) {
    Eigen::Matrix2d M = M_lagrange_2R(q, p);
    Eigen::Vector2d h = h_lagrange_2R(q, dq, p);
    Eigen::Vector2d g = g_lagrange_2R(q, p);
    return M * ddq + h + g;
}

// ------------------------------------------------------------------
// Newton-Euler function to be supplied by you (Lesson 4).
// g_base is the 3D gravity vector in the base frame.
// ------------------------------------------------------------------
Eigen::Vector2d tau_newton_euler_2R(const Eigen::Vector2d& q,
                                    const Eigen::Vector2d& dq,
                                    const Eigen::Vector2d& ddq,
                                    const TwoLinkParams& p,
                                    const Eigen::Vector3d& g_base) {
    // TODO: call your library / implementation here.
    throw std::runtime_error("Connect to your Newton-Euler implementation.");
}

Eigen::Matrix2d reconstruct_M_NE(const Eigen::Vector2d& q,
                                 const TwoLinkParams& p) {
    Eigen::Matrix2d M;
    Eigen::Vector2d dq = Eigen::Vector2d::Zero();
    Eigen::Vector3d g_base(0.0, 0.0, 0.0); // zero gravity

    for (int j = 0; j < 2; ++j) {
        Eigen::Vector2d ddq = Eigen::Vector2d::Zero();
        ddq(j) = 1.0;
        Eigen::Vector2d tau =
            tau_newton_euler_2R(q, dq, ddq, p, g_base);
        M.col(j) = tau;
    }
    return M;
}

Eigen::Vector2d reconstruct_g_NE(const Eigen::Vector2d& q,
                                 const TwoLinkParams& p) {
    Eigen::Vector2d dq  = Eigen::Vector2d::Zero();
    Eigen::Vector2d ddq = Eigen::Vector2d::Zero();
    Eigen::Vector3d g_base(0.0, -p.g, 0.0);
    return tau_newton_euler_2R(q, dq, ddq, p, g_base);
}

int main() {
    TwoLinkParams p;
    p.m1 = 2.0; p.m2 = 1.0;
    p.l1 = 1.0; p.c1 = 0.5; p.c2 = 0.5;
    p.I1 = 0.2; p.I2 = 0.1;
    p.g  = 9.81;

    Eigen::Vector2d q(0.3, -0.7);
    Eigen::Matrix2d M_L  = M_lagrange_2R(q, p);
    Eigen::Matrix2d M_NE = reconstruct_M_NE(q, p);

    std::cout << "M_L(q) =\n" << M_L << std::endl;
    std::cout << "M_NE(q) =\n" << M_NE << std::endl;
    std::cout << "M_L - M_NE =\n" << (M_L - M_NE) << std::endl;

    Eigen::Vector2d g_L  = g_lagrange_2R(q, p);
    Eigen::Vector2d g_NE = reconstruct_g_NE(q, p);
    std::cout << "g_L(q)  = " << g_L.transpose()  << std::endl;
    std::cout << "g_NE(q) = " << g_NE.transpose() << std::endl;

    return 0;
}
      
